src/algorithms/cfd/structures/cfd_relation_data.cpp

```diff
@@ -0,0 +1,295 @@
+                string_row[k] = line[k];
+            }
+            int it;
+            for (size_t i = 0; i < num_columns; i++) {
+                auto ptr = item_dictionary.find(std::make_pair(i, string_row[i]));
+                if (ptr != item_dictionary.end()) {
+                    it = ptr->second;
+                } else {
+                    items.emplace_back(string_row[i], i);
+                    columns_values_dict[static_cast<int>(i)].push_back(unique_elems_number);
+                    item_dictionary[std::make_pair(i, string_row[i])] = unique_elems_number;
+                    it = unique_elems_number++;
+                }
+                int_row[i] = it;
+                items[it - 1].frequency++;
+            }
+            data_rows.push_back(int_row);
+        }
+
+        std::vector<CFDColumnData> column_data;
+        for (int i = 0; static_cast<size_t>(i) < num_columns; ++i) {
+            auto column = Column(schema.get(), parser.GetColumnName(i), i);
+            schema->AppendColumn(std::move(column));
+            column_data.emplace_back(schema->GetColumn(i), columns_values_dict[i]);
+        }
+        schema->Init();
+
+        return std::make_unique<CFDRelationData>(std::move(schema), std::move(column_data),
+                                                 std::move(data_rows), std::move(item_dictionary),
+                                                 std::move(items));
+    }
+}
+
+std::unique_ptr<CFDRelationData> CFDRelationData::CreateFrom(model::IDatasetStream& file_input,
+                                                             bool is_null_eq_null, double c_sample,
+                                                             double r_sample) {
+    // Fields of CFDRelationData class
+    auto schema = std::make_unique<RelationalSchema>(file_input.GetRelationName(), is_null_eq_null);
+    std::vector<Transaction> data_rows;
+    boost::unordered_map<std::pair<int, std::string>, int, PairHash> item_dictionary;
+    std::vector<ItemInfo> items;
+    std::unordered_map<int, std::vector<int>> columns_values_dict;
+    int unique_elems_number = 1;
+    std::random_device rd;   // only used once to initialise (seed) engine
+    std::mt19937 rng(rd());  // random-number engine used (Mersenne-Twister in this case)
+    std::uniform_real_distribution<double> uni(0.0, 1.0);  // guaranteed unbiased
+    int num_columns = static_cast<int>(file_input.GetNumberOfColumns());
+    std::vector<std::string> line;
+    std::vector<int> columns_numbers_list = Range(0, num_columns);
+    int size = static_cast<int>(static_cast<double>(columns_numbers_list.size()) * c_sample);
+    Shuffle(columns_numbers_list);
+    columns_numbers_list =
+            std::vector<int>(columns_numbers_list.begin(), columns_numbers_list.begin() + size);
+    std::sort(columns_numbers_list.begin(), columns_numbers_list.end());
+    while (file_input.HasNextRow()) {
+        if (uni(rng) < r_sample) {
```

Comment:
This method is a bit heavy. Ideally, some parts of it should be extracted into separate methods.

At least, use some `if (...) continue` technique to decrease nestedness.